Document Discord access plugin (#17593)
gravitational/teleport: docs/pages/access-controls/access-request-plugins/ssh-approval-discord.mdx

---
title: Access Requests with Discord
description: How to set up Teleport's Discord plugin for privilege elevation approvals.
---

This guide will explain how to set up Discord to receive Access Request messages
from Teleport. Teleport's Discord integration notifies individuals and channels of
Access Requests. Users can then approve and deny Access Requests from within
Discord, making it easier to implement security best practices without
compromising productivity.

## Prerequisites

(!docs/pages/includes/commercial-prereqs-tabs.mdx!)

- Be admin of the Discord Server (installing a bot requires at least the
  "manager server" permission)
- Either a Linux host or Kubernetes cluster where you will run the Discord plugin.

> (!/docs/pages/includes/tctl.mdx!)

Reviewer: Maybe unnecessary? Seems a bit of a non sequitur.

administrator to grant you a role with the "manage server" permission.

<Admonition type="note">
The same application can be installed into multiple Discord servers. To
do so, access the OAuth URL multiple times and choose different servers. You
have to be admin on a Discord server to install the app into it.
</Admonition>

## Step 6/8. Configure the Teleport Discord plugin

At this point, the Teleport Discord plugin has the credentials it needs to
communicate with your Teleport cluster and the Discord API. In this step, you will
configure the Discord plugin to use these credentials. You will also configure the
plugin to notify the right Discord channels when it receives an Access Request
update.

### Create a config file

<Tabs>
<TabItem label="Executable">
The Teleport Discord plugin uses a config file in TOML format. Generate a
boilerplate config by running the following command (the plugin will not run
unless the config file is in `/etc/teleport-discord.toml`):

```code
$ teleport-discord configure | sudo tee /etc/teleport-discord.toml > /dev/null
```

This should result in a config file like the one below:

```toml

> (!examples/resources/plugins/teleport-discord.toml!)

Reviewer 2: This renders okay for me on the local dev server

> (!examples/resources/plugins/teleport-discord.toml!)

Author: This is a copy-paste from the slack one, so it is likely valid

client_crt = "/var/lib/teleport/plugins/discord/auth.crt" # Teleport GRPC client certificate
root_cas = "/var/lib/teleport/plugins/discord/auth.cas"   # Teleport cluster CA certs
```
</TabItem>
<TabItem label="Connect to the Proxy Service" options="Executable">

Set `addr` to  your Proxy Service address with port `443`.

Set `identity` to the identity file generated earlier:

```toml
[teleport]
addr = "teleport.example.com:443"
identity = "/var/lib/teleport/plugins/discord/auth.pem"
```
</TabItem>
<TabItem label="Connect to the Proxy or Auth Service" scope="cloud" options="Helm Chart">

**`address`**: Include the hostname and port of your Teleport Proxy or Auth Service
(e.g., `teleport.example.com:443`).

**`identitySecretName`**: Fill in the `identitySecretName` field with the name
of the Kubernetes secret you created earlier.

</TabItem>
</Tabs>

</ScopedBlock>
<ScopedBlock scope="cloud">

Set `addr` to  your Teleport Proxy address with port `443`.

Set `identity` to the identity file generated earlier:

```toml
[teleport]
addr = "teleport.example.com:443"
identity = "/var/lib/teleport/plugins/discord/auth.pem"
```

</ScopedBlock>

**`[discord]`**

> `token`: Paste the Bot token saved previously in this field.

Reviewer 2: ```suggestion
`token`: Paste the bot token saved previously in this field.
```

**`[role_to_recipients]`**

> The `role_to_recipients` map configure the channels that the Discord plugin will

Reviewer 2: ```suggestion
The `role_to_recipients` map configures the channels that the Discord plugin will
```

notify when a user requests access to a specific role. When the Discord plugin
receives an Access Request from the Auth Service, it will look up the role being
requested and identify the Discord channels to notify.

> Each channel is represented by a numeric ID, channels can be public, private or

Reviewer 2: ```suggestion
Each channel is represented by a numeric ID. Channels can be public, private or
```

> direct messages between a user and the Bot.

Reviewer 2: ```suggestion
direct messages between a user and the bot.
```

> 

Reviewer 2: ```suggestion
To determine the numeric ID of a channel for the bot to notify, follow the instructions below:

```
To make it more explicit why we're including tabs here

<Tabs>
  <TabItem label="Public Channel">
  

>     Open discord in a web browser and navigate to the desired channel.

Reviewer 2: ```suggestion
    Open Discord in a web browser and navigate to the desired channel.
```

    
    The web browser URL should look like:
    ```
    https://discord.com/channels/<guild ID>/<channel ID>
    ```
    

>     Copy the last part of the URL (everything after the last `/`), this is the channel ID.

Reviewer 2: ```suggestion
    Copy the last part of the URL (everything after the last `/`), which is the channel ID.
```

  </TabItem>
  <TabItem label="Private Channel">

>     Open discord in a web browser and navigate to the desired channel.

Reviewer 2: ```suggestion
    Open Discord in a web browser and navigate to the desired channel.
```

    
    In the channel list choose "Create invite", type "teleport" in the search field

>     and invite your Discord Teleport Bot. The Bot should now appear in the channel

Reviewer 2: ```suggestion
    and invite your Discord Teleport bot. The bot should now appear in the channel
```

    member list.
    
    The web browser URL should look like:
    ```
    https://discord.com/channels/<guild ID>/<channel ID>
    ```
    

>     Copy the last part of the URL (everything after the last `/`), this is the channel ID.

Reviewer 2: ```suggestion
    Copy the last part of the URL (everything after the last `/`), which is the channel ID.
```

  </TabItem>
  <TabItem label="Direct Message">

>     To retrieve the channel ID of the private discussion between UserA and the

Reviewer 2: ```suggestion
    To retrieve the channel ID of the private discussion between User A and the
```

>     Teleport Bot, have UserA send a direct message to the Teleport Bot. This will

Reviewer 2: ```suggestion
    Teleport bot, have User A send a direct message to the Teleport bot. This will
```

>     open a conversation between the user and the Bot. Once the conversation is

Reviewer 2: ```suggestion
    open a conversation between the user and the bot. Once the conversation is
```

    initiated, the user can open the discussion page.
    
    The web browser URL should look like:
    ```
    https://discord.com/channels/@me/<channel ID>
    ```
    

>     Copy the last part of the URL (everything after the last `/`), this is the channel ID.

Reviewer 2: ```suggestion
    Copy the last part of the URL (everything after the last `/`), which is the channel ID.
```
